runtime(filetype): improve *.h filetype detection

selenebun opened the pull request:
This commit changes header files to default to the C filetype, instead of C++, and replaces the existing variables `g:c_syntax_for_h` and `g:ch_syntax_for_h` with a unified `g:filetype_h`, like is used for other ambiguous file extensions.

I preserved[^1] the existing detection of Objective-C files based on the file contents, and also added similar keyword-based detection for C++, which should handle most common situations. I'm sure it could be made more rigorous, but I'm not really much of a C++ programmer.

[^1]: I *did* reduce the number of lines searched to 100 instead of 200, matching most of the other filetypes, but the behavior should otherwise be identical.

Shane-XB-Qian commented on runtime/doc/syntax.txt:
```diff
@@ -1056,8 +1056,6 @@ Variable		Highlight ~
 *c_ansi_typedefs*	 ... but do standard ANSI types
 *c_ansi_constants*	 ... but do standard ANSI constants
 *c_no_utf*		don't highlight \u and \U in strings
-*c_syntax_for_h*	for *.h files use C syntax instead of C++ and use objc
```
you cannot simply just remove `g:c_syntax_for_h`, it is a history.

Shane-XB-Qian reviewed (commented):
you cannot simply just remove `g:c_syntax_for_h`, it is a history.
// BTW: some background you may like to know: https://github.com/vim/vim/pull/13642#issuecomment-1849798999

selenebun committed:
runtime(filetype): improve *.h filetype detection

This commit changes *.h files to default to C (instead of C++), and
deprecates the existing `g:c_syntax_for_h` and `g:ch_syntax_for_h`
variables in favor of a unified `g:filetype_h`, like is used for other
ambiguous file extensions.

Signed-off-by: Amelia Clarke <selene@perilune.dev>

zeertzjq commented on runtime/autoload/dist/ft.vim:
```diff
@@ -203,19 +203,36 @@ export def FTlpc()
+  var n = 1
+  while n < maxlines && n <= line('$')
+    const line = getline(n)
+    if line =~ '\v^\s*\@%(class|interface|end)>'
```
What is the `>` at the end for?

chrisbra replied:
end-of-word ?

zeertzjq replied:
That should be `\>`

chrisbra replied:
not for `\v`
